### PR TITLE
Fix rewrite edge node filter.

### LINKED ISSUE #4989: PredicateExpression `exists(r.likeness)` is wrongly rewritten to `exists((teammate.likeness OR serve.likeness OR like.likeness))`
**Please check the FAQ documentation before raising an issue**

**Describe the bug (__required__)**

PredicateExpression `exists(r.likeness)`  is first embedded into `Traverse` by the rule [`PushFilterDownTraverse`](https://github.com/vesoft-inc/nebula/pull/4987), and then it's wrongly rewritten to `exists((teammate.likeness OR serve.likeness OR like.likeness))`
by the optimizer rule `PushEFilterDown`

**Your Environments (__required__)**

* OS: `uname -a`
* Compiler: `g++ --version` or `clang++ --version`
* CPU: `lscpu`
* Commit id (e.g. `a3ffc7d8`)

**How To Reproduce(__required__)**

```
match (:player{name:"Tony Parker"})-[r]->() where exists(r.likeness) return r, exists({a:12}.a)
```

plan:
```
| name | operator info  |
| Traverse | filter: exists((teammate.likeness OR serve.likeness OR like.likeness)) |
```

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by Shylock-Hg**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
Close https://github.com/vesoft-inc/nebula-ent/issues/1534
Close #4989 

#### Description:
Rewrite edge property like`*.prop` to `_any(e1.prop, e2,prop ...)`

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: tests/tck/features/bugfix/RewriteEdgeFilter.feature [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# This source code is licensed under Apache 2.0 License.
+Feature: Test match used in pipe
+
+  Background:
+    Given a graph with space named "nba"
+
+  Scenario: Order by after match
+    When executing query:
+      """
+      match (v)-[e:like|teammate{start_year: 2010}]->() where id(v) == 'Tim Duncan' return e
```
Comment (yixinglu) [inline tests/tck/features/bugfix/RewriteEdgeFilter.feature:12]: In openCypher, this edge filter means to find these edges which contain like or teammate labels and have the property start_year equal to 2010. So I think this filter should be converted to following format in NebulaGraph:

```
("like" in tags(e) and e.like.start_year = 2010) or ("teammate" in tags(e) and e.teammate.start_year=2010)
```

Comment (Shylock-Hg) [reply tests/tck/features/bugfix/RewriteEdgeFilter.feature:12]: This is not only `*.prop = xxxx`, but we should have an equivalent representation of `*.prop` of edge property consists multiple edge types.

File: src/common/function/FunctionManager.cpp [hunk 2]
```diff
@@ -41,6 +42,7 @@ std::unordered_map<std::string, Value::Type> FunctionManager::variadicFunReturnT
     {"concat_ws", Value::Type::STRING},
     {"cos_similarity", Value::Type::FLOAT},
     {"coalesce", Value::Type::__EMPTY__},
+    {"_any", Value::Type::__EMPTY__},
```
Comment (codesigner) [inline src/common/function/FunctionManager.cpp:45]: the new added _any function, is it callable from query?

Comment (Shylock-Hg) [reply src/common/function/FunctionManager.cpp:45]: Can't.

File: src/graph/optimizer/rule/PushEFilterDownRule.cpp [hunk 3]
```diff
@@ -123,17 +124,15 @@ std::string PushEFilterDownRule::toString() const {
...
       for (auto &edge : edges) {
         auto reEdgeExp = rewriteStarEdge(propertyExpr, spaceId, edge, schemaMng, pool);
         if (reEdgeExp == nullptr) {
           return nullptr;
         }
-        operands.emplace_back(reEdgeExp);
+        args->addArgument(reEdgeExp);
       }
-      static_cast<LogicalExpression *>(ret)->setOperands(std::move(operands));
+      ret = FunctionCallExpression::make(pool, "_any", args);
```
Comment (czpmango) [inline src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: Why not just rewrite it as `like.start_year==2010 or teammate.start_year==2010`?

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: How about others?

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: `like.start_year > xx' ?

Comment (yixinglu) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: > `like.start_year > xx' ?

This filter could not be used in path pattern. I don't understand the purpose of the `_any` inner function.

Comment (Shylock-Hg) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: Storage don't support `*.prop`, so rewrite it.

File: src/common/function/FunctionManager.cpp [hunk 4]
```diff
@@ -2799,6 +2801,21 @@ FunctionManager::FunctionManager() {
...
+    attr.minArity_ = 1;
+    attr.maxArity_ = INT64_MAX;
+    attr.isAlwaysPure_ = true;
+    attr.body_ = [](const auto &args) -> Value {
+      for (const auto &arg : args) {
+        if (!arg.get().isNull() && !arg.get().empty()) {
+          return arg.get();
+        }
+      }
+      return Value::kNullBadData;
```
Comment (czpmango) [inline src/common/function/FunctionManager.cpp:2816]: `kNullBadData` is designed for error and`kNullValue` is sufficient for ignoring behavior. Fwiw, the `bad null` should be passed through by all functions or expressions.

Comment (Shylock-Hg) [reply src/common/function/FunctionManager.cpp:2816]: What' your meaning?

File: src/graph/optimizer/rule/PushEFilterDownRule.cpp (hunk 3 above)
Comment (yixinglu) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: Agree with @czpmango comment.

**Review by xtcyclist — COMMENTED** [review]
`*.prop` equals to something like `union(e1.prop, e2,prop ...)` semantically, not `_any(e1.prop, e2,prop ...)`. Please make sure it's ok to change such semantics.

**Comment by Shylock-Hg** [comment]
> `*.prop` equals to something like `union(e1.prop, e2,prop ...)` semantically, not `_any(e1.prop, e2,prop ...)`. Please make sure it's ok to change such semantics.

`match (v)-[e:like|serve]->() return e.likeness' will return any nonull value of `like. Likeness` or `serve.likeness`

File: src/common/function/FunctionManager.cpp (hunk 2 above)
Comment (xtcyclist) [reply src/common/function/FunctionManager.cpp:45]: If I understood correctly, it seems to me that you mean `_all`, not `_any`. If there are 10 edge types, all of which has the same prop: `prop_x`, the semantics of `any` could be satisfied if only 1 such `prop_x` is found from all 10 edge types. `all` means making sure and finding all `prop_x` from all 10 edge types. I think you mean the latter one.

File: src/graph/optimizer/rule/PushEFilterDownRule.cpp (hunk 3 above)
Comment (xtcyclist) [reply src/graph/optimizer/rule/PushEFilterDownRule.cpp:135]: I think what @Shylock-Hg wants is a generic way. In the path pattern alone, however, rewriting with == suffices.

**Review by xtcyclist — APPROVED** [review]

**Comment by jievince** [comment]
Please look at this issue: https://github.com/vesoft-inc/nebula/issues/4989

File: src/common/function/FunctionManager.cpp [hunk 5]
```diff
@@ -2820,6 +2822,21 @@ FunctionManager::FunctionManager() {
...
+    attr.isAlwaysPure_ = true;
+    attr.body_ = [](const auto &args) -> Value {
+      for (const auto &arg : args) {
+        if (!arg.get().isNull() && !arg.get().empty()) {
+          return arg.get();
+        }
+      }
+      return Value::kNullValue;
+    };
+  }
```
[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/function/FunctionManager.cpp:2839]: ok

[the remark this answers is not in the document]
Comment (jievince) [reply src/common/function/FunctionManager.cpp:2839]: Better to resuse it.

[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/function/FunctionManager.cpp:2839]: We should ignore `EMPTY` too.

**Review by jievince — APPROVED** [review]

**Pull request merged**